google/cloud/forseti/services/inventory/storage.py

```diff
@@ -937,217 +906,50 @@ def write(self, resource):
 
         Args:
             resource (object): Resource object to store in db.
-
-        Raises:
-            Exception: If storage was opened readonly.
         """
-        if self.readonly:
-            raise Exception('Opened storage readonly')
-
-        resource_data = resource.data()
-        # Group members do not need to be checked if it already exists
-        # in Inventory, as a user can be members in multiple groups.
-        if resource_data.get('kind') != 'admin#directory#member':
-            previous_id = self._get_resource_id(resource)
-            if previous_id:
-                resource.set_inventory_key(previous_id)
-                self.update(resource)
+        # Use a lock to quickly check if this is a duplicate resource before
+        # updating the cache and proceeding.
+        with self._storage_lock:
+            if resource.get_full_resource_name() in self._wrote_resources:
+                LOGGER.warning('Duplicate Resource in inventory, skipping %s',
+                               resource.get_full_resource_name())
                 return
+            self._wrote_resources.add(resource.get_full_resource_name())
 
-        rows = Inventory.from_resource(self.inventory_index, resource)
-
-        for row in rows:
-            if row.category == Categories.resource:
-                # Force flush to insert the resource row in order to get the
-                # inventory id value. This is used to tie child resources
-                # and related data back to the parent resource row and to
-                # check for duplicate resources.
-                self.session.add(row)
-                self.session.flush()
-                resource.set_inventory_key(row.id)
-            else:
-                row.parent_id = resource.inventory_key()
-                self.buffer.add(row)
-
-        self.inventory_index.counter += len(rows)
-
-    def update(self, resource):
-        """Update a resource in the storage.
-
-        Args:
-            resource (object): Resource object to store in db.
-
-        Raises:
-            Exception: If storage was opened readonly.
-        """
+        (resource_row, policy_rows) = Inventory.from_resource(
+            self.inventory_index, resource)
 
-        if self.readonly:
-            raise Exception('Opened storage readonly')
+        # Insert first row to get the primary key for the resource
```

Comment:
if we leverage the relationship pattern in sqlalchemy we can do bulk insert without worrying about the primary key / foreign key reference, what do you think about that approach?

Note: I do not know about the performance about this approach

https://docs.sqlalchemy.org/en/13/orm/basic_relationships.html

Reply:
I expect behind the scenes the performance is the same or worse, as it has to do the same set of operations in order to create the row, get the id, and the use the id to create the other rows. This way we are explicit. And I'd like to keep from having to maintain a session in each thread in order to do inserts.